tls/bignum.c

```diff
@@ -1,1009 +1,865 @@
+ */
+int
+ttls_mpi_fill_random(TlsMpi *X, size_t size)
+{
+	size_t limbs = CHARS_TO_LIMBS(size);
 
-	*p++ = '\0';
-	*olen = p - buf;
+	if (WARN_ON_ONCE(size > TTLS_MPI_MAX_SIZE))
+		return -EINVAL;
 
-cleanup:
+	if (__mpi_realloc(X, limbs, 0))
+		return -ENOMEM;
 
-	ttls_mpi_free(&T);
+	ttls_rnd(X->p, size);
+	X->used = limbs;
+	X->s = 1;
 
-	return ret;
+	return 0;
 }
 
-/*
- * Import X from unsigned binary data, big endian
+/**
+ * Compare unsigned values.
  */
-int ttls_mpi_read_binary(ttls_mpi *X, const unsigned char *buf, size_t buflen)
+int
+ttls_mpi_cmp_abs(const TlsMpi *X, const TlsMpi *Y)
 {
-	int ret;
-	size_t i, j;
-	size_t const limbs = CHARS_TO_LIMBS(buflen);
-
-	/* Ensure that target MPI has exactly the necessary number of limbs */
-	if (X->n != limbs)
-	{
-		ttls_mpi_free(X);
-		ttls_mpi_init(X);
-		TTLS_MPI_CHK(ttls_mpi_grow(X, limbs));
-	}
+	int i;
 
-	TTLS_MPI_CHK(ttls_mpi_lset(X, 0));
+	if (!X->used && !Y->used)
+		return 0;
 
-	for (i = buflen, j = 0; i > 0; i--, j++)
-		X->p[j / ciL] |= ((ttls_mpi_uint) buf[i - 1]) << ((j % ciL) << 3);
+	if (X->used > Y->used)
+		return 1;
+	if (Y->used > X->used)
+		return -1;
 
-cleanup:
+	for (i = X->used - 1; i >= 0; i--) {
+		if (X->p[i] == Y->p[i])
+			continue;
+		return X->p[i] > Y->p[i] ? 1 : -1;
+	}
 
-	return ret;
+	return 0;
 }
 
 /*
- * Export X into unsigned binary data, big endian
+ * Compare signed values.
  */
-int ttls_mpi_write_binary(const ttls_mpi *X, unsigned char *buf, size_t buflen)
+int
+ttls_mpi_cmp_mpi(const TlsMpi *X, const TlsMpi *Y)
 {
-	size_t i, j, n;
+	int i;
 
-	n = ttls_mpi_size(X);
+	if (!X->used && !Y->used)
+		return 0;
 
-	if (buflen < n)
-		return(TTLS_ERR_MPI_BUFFER_TOO_SMALL);
+	if (X->used > Y->used)
+		return X->s;
+	if (Y->used > X->used)
+		return -Y->s;
 
-	memset(buf, 0, buflen);
+	if (X->s > 0 && Y->s < 0)
+		return 1;
+	if (Y->s > 0 && X->s < 0)
+		return -1;
 
-	for (i = buflen - 1, j = 0; n > 0; i--, j++, n--)
-		buf[i] = (unsigned char)(X->p[j / ciL] >> ((j % ciL) << 3));
+	for (i = X->used - 1; i >= 0; i--) {
+		if (X->p[i] == Y->p[i])
+			continue;
+		return X->p[i] > Y->p[i] ? X->s : -X->s;
+	}
 
 	return 0;
 }
 
-/*
- * Left-shift: X <<= count
+/**
+ * Compare MPI with a signed value.
  */
-int ttls_mpi_shift_l(ttls_mpi *X, size_t count)
+int
+ttls_mpi_cmp_int(const TlsMpi *X, long z)
 {
-	int ret;
-	size_t i, v0, t1;
-	ttls_mpi_uint r0 = 0, r1;
-
-	v0 = count / (biL	);
-	t1 = count & (biL - 1);
+	if (X->used > 1)
+		return X->s;
+	if (!X->used)
+		return z == 0 ? 0 : z < 0 ? 1 : -1;
+
+	if (z < 0) {
+		if (X->s > 0)
+			return 1;
+		z = -z;
+	} else {
+		if (X->s < 0)
+			return -1;
+	}
 
-	i = ttls_mpi_bitlen(X) + count;
+	/* Modular comparison. */
+	return z == X->p[0] ? 0 : X->p[0] > z ? X->s : -X->s;
+}
 
-	if (X->n * biL < i)
-		TTLS_MPI_CHK(ttls_mpi_grow(X, BITS_TO_LIMBS(i)));
+/**
+ * Unsigned addition: X = |A| + |B|
+ *
+ * @A and @B must be different, but either of them can accept the result @X.
```

Comment:
This may be a neat interface for a general purpose library, but that flexibility usually is not needed, as one knows whenever they do addition `X = A + B` where all numbers are in different memory blocks, or accumulation `A += B'. Flexibility makes code slower and more complex. I'd prefer to have two separate functions for the two cases.